### ISSUE TITLE
Proxy: containers stuck in "CreateContainer" phase for peer Pods

### ISSUE THREAD

**Issue by RedCarsten** [confidential-containers/cloud-api-adaptor#1402]
When using peerPod deployments with `>1` we _sometimes_ see that pods would be stuck in "CreateContainer" phase, so the deployment never reaches its configured number of replicas. We always say this with an odd number of replicas, but that might be coincidence. 
The number of peer pod VMs matches the number of replicas, but the create container phase does not complete.

We found that the cause is in https://github.com/confidential-containers/cloud-api-adaptor/blob/main/pkg/adaptor/proxy/proxy.go#L121

```go
err := retry.Do(
		func() error {
			var err error
			conn, err = dialer.DialContext(ctx, "tcp", address)
			return err
		},
		retry.Attempts(0),
		retry.Context(ctx),
		retry.MaxDelay(5*time.Second),
	)
```

The `conn, err = dialer.DialContext(ctx, "tcp", address)` continuously fails in the error case. After adding some debug logs we found as cause:

```text
2023/09/05 10:41:53 [adaptor/proxy] Carsten agentProxy DialContext [192.168.122.179:15150], cause: tls: failed to verify certificate: x509: certificate is valid for podvm-dep-64043-rc49d5my-54b5bc4c7f-4qphd-acbaef60, not podvm-dep-64043-rc49d5my-54b5bc4c7f-rnwzd-7964b546
2023/09/05 10:41:53 Carsten Exit  [agentProxy]: [DialContext: 192.168.122.179:15150, /run/peerpod/pods/7964b5461b2356b9f64ca0bd78250ca7a1249661e48ca444d114862b90ba96b6/agent.ttrpc] in [1.83119ms]
```

In this log snippet
```
podvm-dep-64043-rc49d5my-54b5bc4c7f-4qphd-acbaef60
```

is one pod of the replica set that has been created successfully. 

```
podvm-dep-64043-rc49d5my-54b5bc4c7f-rnwzd-7964b546
```

is the pod that could not be created successfully.

Without understanding all details it looks as if there exists some sort of race condition that causes a mix-up of certificates between pods.

**Comment by RedCarsten** [comment]
[bad.log](https://github.com/confidential-containers/cloud-api-adaptor/files/12523030/bad.log)

This log shows the error scenario inlcuding a number of debug patches (starting by `Carsten`). The ID of the failing pod is `podvm-dep-64043-rc49d5my-54b5bc4c7f-rnwzd-7964b546`.

**Comment by RedCarsten** [comment]
cc @bpradipt @yoheiueda

**Comment by yoheiueda** [comment]
It looks like a race condition...

It is hard to follow the log since messages for different VMs are interleaving. Could you modify the libvirt provider to show VM name or SID in debug message for IP address?

```
2023/09/05 10:38:21 [adaptor/cloud/libvirt] VM id 474
2023/09/05 10:38:21 [adaptor/cloud/libvirt] VM id 475
2023/09/05 10:38:21 [adaptor/cloud/libvirt] VM id 476
2023/09/05 10:38:21 [adaptor/cloud/libvirt] VM id 477
2023/09/05 10:38:22 [adaptor/cloud/libvirt] 380928 bytes uploaded
2023/09/05 10:38:22 [adaptor/cloud/libvirt] Volume ID: /var/lib/libvirt/images/podvms/podvm-dep-64043-rc49d5my-54b5bc4c7f-4qphd-acbaef60-cloudinit.iso
2023/09/05 10:38:22 [adaptor/cloud/libvirt] Create XML for 'podvm-dep-64043-rc49d5my-54b5bc4c7f-4qphd-acbaef60'
2023/09/05 10:38:22 [adaptor/cloud/libvirt] Creating VM 'podvm-dep-64043-rc49d5my-54b5bc4c7f-4qphd-acbaef60'
2023/09/05 10:38:22 [adaptor/cloud/libvirt] Starting VM 'podvm-dep-64043-rc49d5my-54b5bc4c7f-4qphd-acbaef60'
2023/09/05 10:38:22 [adaptor/cloud/libvirt] VM id 478
2023/09/05 10:38:51 [adaptor/cloud/libvirt] domain IP details [{vnet406 52:54:00:0a:27:9e [{0 192.168.122.179 24}]}]
2023/09/05 10:38:51 [adaptor/cloud/libvirt] VM IP 192.168.122.179
2023/09/05 10:38:51 [adaptor/cloud/libvirt] VM IP list [192.168.122.179]
```

**Comment by CarstenLeue** [comment]
Looks like the race condition is here:

https://github.com/confidential-containers/cloud-api-adaptor/blob/main/pkg/adaptor/cloud/libvirt/libvirt.go#L65

if `createCloudInitISO` gets called concurrently for different pods, there is a chance that the `user-data` file gets overridden, so the resulting ISO file would contain incorrect information.

This is fixed as a side effect by https://github.com/confidential-containers/cloud-api-adaptor/pull/1268

cc @bpradipt

**Comment by RedCarsten** [comment]
#1268 has been merged
